Resources/Storage/GFAL2_SRM2Storage.py

```diff
@@ -97,77 +96,121 @@ def getTransportURL( self, path, protocols = False ):
 
-
-  def __getSingleTransportURL( self, path, protocols = False ):
+  def __getSingleTransportURL(self, path, protocols=False):
     """ Get the tURL from path with getxattr from gfal2
 
     :param self: self reference
     :param str path: path on the storage
     :returns: S_OK( Transport_URL ) in case of success
               S_ERROR( errStr ) in case of a failure
     """
-    self.log.debug( 'GFAL2_SRM2Storage.__getSingleTransportURL: trying to retrieve tURL for %s' % path )
+    self.log.debug(
+        'GFAL2_SRM2Storage.__getSingleTransportURL: trying to retrieve tURL for %s' %
+        path)
     if protocols:
-      self.ctx.set_opt_string_list( "SRM PLUGIN", "TURL_PROTOCOLS", protocols )
+      self.ctx.set_opt_string_list("SRM PLUGIN", "TURL_PROTOCOLS", protocols)
 
-    res = self._getExtendedAttributes( path, attributes = ['user.replicas'] )
+    res = self._getExtendedAttributes(path, attributes=['user.replicas'])
     self.__setSRMOptionsToDefault()
 
     if res['OK']:
-      return S_OK( res['Value']['user.replicas'] )
-    else:
-      errStr = 'GFAL2_SRM2Storage.__getSingleTransportURL: Extended attribute tURL is not set.'
-      self.log.debug( errStr, res['Message'] )
-      return res
+      return S_OK(res['Value']['user.replicas'])
+
+    errStr = 'GFAL2_SRM2Storage.__getSingleTransportURL: Extended attribute tURL is not set.'
+    self.log.debug(errStr, res['Message'])
+    return res
+
+  def getOccupancy(self, *parms, **kws):
+    """ Gets the GFAL2_SRM2Storage occupancy info.
+
+      TODO: needs gfal2.15 because of bugs:
+      https://its.cern.ch/jira/browse/DMC-979
+      https://its.cern.ch/jira/browse/DMC-977
+
+      It queries the srm interface for a given space token.
+      Out of the results, we keep totalsize, guaranteedsize, and unusedsize all in MB.
+    """
+
+    # Gfal2 extended parameter name to query the space token occupancy
+    spaceTokenAttr = 'spacetoken.description?%s' % self.protocolParameters['SpaceToken']
+    # gfal2 can take any srm url as a base.
+    spaceTokenEndpoint = self.getURLBase()['Value']
+    print "%s %s" % (spaceTokenEndpoint, spaceTokenAttr)
+    try:
+      occupancyStr = self.ctx.getxattr(spaceTokenEndpoint, spaceTokenAttr)
+      try:
+        occupancyDict = json.loads(occupancyStr)
+      except ValueError:
+        # https://its.cern.ch/jira/browse/DMC-977
+        # a closing bracket is missing, so we retry after adding it
+        occupancyStr = occupancyStr[:-1] + '}]'
+        occupancyDict = json.loads(occupancyStr)[0]
+
+        # https://its.cern.ch/jira/browse/DMC-979
+        # We set totalsize to guaranteed size
+        # (it is anyway true for all the SEs I could test)
+        occupancyDict['totalsize'] = occupancyDict.get('guaranteedsize', 0)
+
+    except (gfal2.GError, ValueError) as e:
+      errStr = 'Something went wrong while checking for spacetoken occupancy.'
+      self.log.verbose(errStr, e.message)
```

Comment:
`verbose`? or `error`?

Reply:
I'd say `verbose`, because the error is returned to the StorageElement which will try another plugin